cmd/buckets-scan-mgr.go

```diff
@@ -0,0 +1,246 @@
+				buckets, err := mgr.bucketsLister(mgr.ctx, BucketOptions{})
+				if err == nil {
+					mgr.mu.Lock()
+					mgr.knownBuckets = make(map[string]struct{}, len(buckets))
+					for _, b := range buckets {
+						mgr.knownBuckets[b.Name] = struct{}{}
+					}
+					for bucket := range mgr.knownBuckets {
+						for _, set := range mgr.internal {
+							st := set[bucket]
+							if l, err := globalLifecycleSys.Get(bucket); err == nil && l.HasActiveRules("") {
+								st.lifecycle = true
+							}
+							set[bucket] = st
+						}
+					}
+					mgr.mu.Unlock()
+				}
+				t.Reset(bucketsListInterval)
+			case <-mgr.ctx.Done():
+				return
+			}
+		}
+	}()
+
+	// Clean up internal data when a deleted bucket is found
+	go func() {
+		const cleanInterval = 30 * time.Second
+
+		t := time.NewTimer(cleanInterval)
+		defer t.Stop()
+
+		for {
+			select {
+			case <-t.C:
+				mgr.mu.Lock()
+				for _, set := range mgr.internal {
+					for bkt := range set {
+						if _, ok := mgr.knownBuckets[bkt]; !ok {
+							delete(set, bkt)
+						}
+					}
+				}
+				mgr.mu.Unlock()
+
+				t.Reset(cleanInterval)
+			case <-mgr.ctx.Done():
+				return
+			}
+		}
+	}()
+
+	// A routine that sends the next bucket to scan for each erasure set listener
+	go func() {
+		tick := 10 * time.Second
+
+		t := time.NewTimer(tick)
+		defer t.Stop()
+
+		for {
+			select {
+			case <-t.C:
+				mgr.mu.RLock()
+				for id, ch := range mgr.bucketsCh {
+					if len(ch) == 0 {
+						b := mgr.unsafeGetNextBucket(id)
+						if b != "" {
+							select {
+							case ch <- b:
+							default:
+							}
+						}
+					}
+				}
+				mgr.mu.RUnlock()
+
+				t.Reset(tick)
+			case <-mgr.ctx.Done():
+				return
+			}
+		}
+	}()
+}
+
+// Return a channel of buckets names to scan a given erasure set identifier
+func (mgr *bucketsScanMgr) getBucketCh(id int) chan string {
+	mgr.mu.Lock()
+	defer mgr.mu.Unlock()
+
+	mgr.internal[id] = make(map[string]bucketScanStat)
+	mgr.bucketsCh[id] = make(chan string, 1)
+
+	return mgr.bucketsCh[id]
+}
+
+func less(st1, st2 bucketScanStat) bool {
+	if st1.ongoing != st2.ongoing {
+		return st1.ongoing == false
+	}
+	if st1.cycle != st2.cycle {
+		return st1.cycle < st2.cycle
+	}
+	if st1.lifecycle != st2.lifecycle {
+		return st1.lifecycle == true
+	}
+	return false
+}
+
+// Return the next bucket name to scan of a given erasure set identifier
+// If all buckets are in a scanning state, return empty result
+func (mgr *bucketsScanMgr) unsafeGetNextBucket(id int) string {
+	var (
+		nextBucketStat = bucketScanStat{}
+		nextBucketName = ""
+	)
+
+	for bucket, stat := range mgr.internal[id] {
+		if stat.ongoing {
+			continue
+		}
+		if nextBucketName == "" {
+			nextBucketName = bucket
+			nextBucketStat = stat
+			continue
+		}
+		if nextBucketName == "" || less(stat, nextBucketStat) {
+			nextBucketStat = stat
+			nextBucketName = bucket
+		}
+	}
+
+	return nextBucketName
+}
+
+// Mark a bucket as done in a specific erasure set - returns true if successful,
```

Comment:
`%s/Mark a bucket as done in a specific erasure set/Mark a bucket as scan started in a specific erasure set/g`